### PR TITLE
Switch ovn-k8s-overlay to golang

### PR DESCRIPTION
**Pull request by mozhuli**
Switch ovn-k8s-overlay to golang

NOTE: Testing on redhat and windows has not yet completed.

@shettyg @feiskyer

### CODE REVIEW

**Comment by mozhuli** [comment]
There are some works have not completed. eg. RedHat and win32 os enviroment and advanced config values etc. And i am working on it. so I change the  PR to WIP. 
@shettyg

**Comment by shettyg** [comment]
While at it, can you also please look at the linter failures. Just some styling issues.

File: go-controller/cmd/ovn-k8s-overlay/app/common.go [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+const (
+	// ConfigFilePath is the config file path for more fine grained control..
+	ConfigFilePath = "/etc/openvswitch/ovn_k8s.conf"
+)
+
+var cfg config.Config
+
+func fetchOVNNB(context *cli.Context) error {
+	// Open ovn_k8s config file.
+	conf, err := os.Open(ConfigFilePath)
```
Comment (shettyg) [inline go-controller/cmd/ovn-k8s-overlay/app/common.go]: We may as well have a fetchConfig function called from initMaster. It is a little confusing to get all these values from a function named "fetchOVNNB"

File: go-controller/cmd/ovn-k8s-overlay/app/common.go [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+	// Make sure br-int is created
+	stdout, stderr, err = util.RunOVSVsctl("--", "--may-exist", "add-br", "br-int")
+	if err != nil {
+		logrus.Errorf("Failed to create br-int, stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+		return err
+	}
+
+	// Create a OVS internal interface
+	interfaceName := "k8s-" + (nodeName[:11])
+	stdout, stderr, err = util.RunOVSVsctl("--", "--may-exist", "add-port", "br-int", interfaceName, "--", "set", "interface", interfaceName, "type=internal", "mtu_request=1400", "external-ids:iface-id=k8s-"+nodeName)
```
Comment (shettyg) [inline go-controller/cmd/ovn-k8s-overlay/app/common.go]: The MTU should be gotten from the config. The python code does so.

File: go-controller/cmd/ovn-k8s-overlay/app/minion_init.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	if clusterIPSubnet == "" {
+		return fmt.Errorf("argument --cluster-ip-subnet should be non-null")
+	}
+
+	nodeName := context.String("node-name")
+	if nodeName == "" {
+		return fmt.Errorf("argument --node-name should be non-null")
+	}
+
+	cniPluginPath, err := exec.LookPath(config.CniPlugin)
```
Comment (shettyg) [inline go-controller/cmd/ovn-k8s-overlay/app/minion_init.go]: The lines from here for CNI configuration are only run for Linux (skipped for windows). Can you just do the same  here too.

**Comment by shettyg** [comment]
Can you please add the govendor commands that you used for vendoring to the commit log. It will help with rebasing in the future. 

I did not test this, and just went through the file to see the general structure. So few review comments.

**Comment by mozhuli** [comment]
@shettyg  Updated

File: go-controller/cmd/ovn-k8s-overlay/app/common.go (hunk 1 above)
Comment (mozhuli) [reply go-controller/cmd/ovn-k8s-overlay/app/common.go]: add fetchConfig function

File: go-controller/cmd/ovn-k8s-overlay/app/common.go [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+	// ConfigFilePath is the config file path for more fine grained control..
+	ConfigFilePath = "/etc/openvswitch/ovn_k8s.conf"
+)
+
+var cfg config.Config
+
+func fetchOVNNB(context *cli.Context) error {
+	// Open ovn_k8s config file.
+	conf, err := os.Open(ConfigFilePath)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (mozhuli) [reply go-controller/cmd/ovn-k8s-overlay/app/common.go]: fixed

File: go-controller/cmd/ovn-k8s-overlay/app/common.go (hunk 2 above)
Comment (mozhuli) [reply go-controller/cmd/ovn-k8s-overlay/app/common.go]: updated

File: go-controller/cmd/ovn-k8s-overlay/app/minion_init.go (hunk 3 above)
Comment (mozhuli) [reply go-controller/cmd/ovn-k8s-overlay/app/minion_init.go]: updated

**Comment by shettyg** [comment]
Looks like in RunOVNNbctl, we are getting config.NbctlPrivateKey which will fail if you don't have a config file. Also, in gateway_init, you will don't seem to be calling config.FetchConfig. Can you run these once without the config file in the system?

**Comment by mozhuli** [comment]
@shettyg Update, we can simplify config init when using two executables eventually.

**Pull request merged**